### DISCUSSION TITLE
Utilizzo del Token Personale

### DISCUSSION THREAD

**Discussion by dmigliacci** [fattureincloud/api#105] [Q&A] [answered]
Buongiorno,
Sto iniziando il porting dall'integrazione V1 alla V2 e ci sono un po' di quesiti da smarcare...

Fo generato un "token personale" assegnato ad una specifica app tramite il **portale -> impostazioni -> sviluppatore.**

Negli esempi e nella guida però trovo sempre Oauth2, con Authorization Code + Access + Refresh.

**È possibile utilizzare il token personale nell'applicazione per fare le chiamate tramite la vostra SDK?**

Ho utilizzato questo tipo di autenticazione trovato nelle guide, ma va in errore già al **getDefaultApiClient**:

```
ApiClient defaultClient = Configuration.getDefaultApiClient();
//set your access token
OAuth OAuth2AuthenticationCodeFlow = (OAuth) defaultClient.getAuthentication("OAuth2AuthenticationCodeFlow");
OAuth2AuthenticationCodeFlow.setAccessToken("YOUR ACCESS TOKEN");
```

Grazie in anticipo,
Davide

**Comment by valmoz** [comment, marked as answer]
Buongiorno @dmigliacci,
sì, è certamente possibile utilizzare le SDK tramite Token Manuale, in questo caso però se l'errore è su getDefaultApiClient il problema non è legato al token ma ad altro.
Per favore, ci invii il messaggio di errore ed ulteriori informazioni che crede potrebbero essere utili, in caso contrario diventa difficile fornirle supporto.

Grazie 
Mauro

**Comment by dmigliacci** [comment]
Ciao @valmoz,
Proprio alla chiamata:

`ApiClient defaultClient = Configuration.getDefaultApiClient();
`
Ricevo il seguente stackTrace:

`
**java.lang.NoSuchMethodError: kotlin.collections.ArraysKt.copyInto([B[BIII)[B**

        at okio.Segment.writeTo(Segment.kt:169)
        at okio.Segment.compact(Segment.kt:152)
        at okio.Buffer.write(Buffer.kt:1842)
        at okio.Buffer.read(Buffer.kt:1854)
        at okio.Buffer.writeAll(Buffer.kt:1642)
        at okio.Options$Companion.buildTrieRecursive(Options.kt:187)
        at okio.Options$Companion.buildTrieRecursive(Options.kt:174)
        at okio.Options$Companion.buildTrieRecursive$default(Options.kt:113)
        at okio.Options$Companion.of(Options.kt:72)
        at okhttp3.internal.Util.<clinit>(Util.kt:70)
        at okhttp3.internal.concurrent.TaskRunner.<clinit>(TaskRunner.kt:309)
        at okhttp3.ConnectionPool.<init>(ConnectionPool.kt:41)
        at okhttp3.ConnectionPool.<init>(ConnectionPool.kt:47)
        at okhttp3.OkHttpClient$Builder.<init>(OkHttpClient.kt:471)
        at it.fattureincloud.sdk.ApiClient.initHttpClient(ApiClient.java:109)
        at it.fattureincloud.sdk.ApiClient.initHttpClient(ApiClient.java:105)
        at it.fattureincloud.sdk.ApiClient.<init>(ApiClient.java:80)
        at it.fattureincloud.sdk.Configuration.<clinit>(Configuration.java:19)
        at xxxxxxxxxxxFattureInCloudServiceV2Impl.newProformaV2(FattureInCloudServiceV2Impl.java:65)
        at xxxxxxxxxxxFattureInCloudServiceV2Impl$$FastClassBySpringCGLIB$$718b1a4a.invoke(<generated>)`

Grazie per la rapida risposta, resto in attesa.
Davide

**Reply by valmoz** [reply]
Questo è un errore che sembra verificarsi su una delle dipendenze della nostra libreria (cioé OkHttp).
In particolare, sul metodo: kotlin.collections.ArraysKt.copyInto([B[BIII)[B

le chiederei di controllare la sua versione di Java e le relative dipendenze, come consigliato per un [problema simile sulle SDK di Slack](https://github.com/slackapi/java-slack-sdk/issues/518).

I requisiti di OkHttp: https://square.github.io/okhttp/#requirements

Grazie
Mauro

**Reply by dmigliacci** [reply]
Ok grazie, approfondisco e ti faccio sapere. Che versione di Java minima è richiesta per la vostra sdk?

**Reply by emajo** [reply]
Ciao @dmigliacci la versione minima richiesta dalla nostra SDK è java 8

**Reply by valmoz** [reply]
Ciao @dmigliacci esatto, come dice @emajo la versione minima è Java 8, che corrisponde anche alla versione minima per OkHttp.
Puoi trovare questa informazione nella [Repository GitHub](https://github.com/fattureincloud/fattureincloud-java-sdk).

Nel caso in cui la tua versione Java dovesse essere inferiore, puoi sempre optare per lo sviluppo manuale delle chiamate alle API, utilizzando un client Http. Ovviamente questa soluzione richiede uno sforzo maggiore da parte tua, in quanto non avrai a disposizione i modelli offerti dalla SDK.

Mauro

**Reply by dmigliacci** [reply]
Ciao, la versione java è la 1.8 quindi ci siamo, controllando le dependencies di gradle la versione di Okhtto è la **com.squareup.okhttp3:okhttp:4.9.3**

Probabilmete l'errore è nella libreria Kotlin?

```
it.fattureincloud:fattureincloud-java-sdk:2.+ -> 2.0.12
|    +--- io.swagger:swagger-annotations:1.6.6
|    +--- com.google.code.findbugs:jsr305:3.0.2
|    +--- com.squareup.okhttp3:okhttp:4.9.3
|    |    +--- com.squareup.okio:okio:2.8.0
|    |    |    +--- org.jetbrains.kotlin:kotlin-stdlib:1.4.0 -> 1.2.41
|    |    |    |    \--- org.jetbrains:annotations:13.0
|    |    |    \--- org.jetbrains.kotlin:kotlin-stdlib-common:1.4.0
|    |    \--- **org.jetbrains.kotlin:kotlin-stdlib:1.4.10 -> 1.2.41 (*)**
```

**Reply by valmoz** [reply]
Ciao @dmigliacci 
esatto, come può evincere dai log la libreria dipende da kotlin 1.3

Può provare a forzare la dipendenza kotlin come segue:
```
<dependency>
    <groupId>org.jetbrains.kotlin</groupId>
    <artifactId>kotlin-stdlib</artifactId>
    <version>1.3.70</version>
</dependency>
```

**Reply by dmigliacci** [reply]
Ciao @valmoz,
posso confermare che aggiornando la libreria Kotlin 

`implementation group: 'org.jetbrains.kotlin', name: 'kotlin-stdlib', version: '1.7.10'`

Tutta la chiamata fila liscia, devo solo verificare che non ci siano controindicazioni sul resto.

Vi ringrazio per supporto rapido e super accurato, procedo nella migrazione e ci sentiremo su altri ticket ;-)
